### ISSUE TITLE
Changes not saved when editing workload deployment config

### ISSUE THREAD

**Issue by brudnak** [rancher/dashboard#4398]
**Rancher Server Setup**
- Rancher version: v2.6-a636bda2a00e939665b381f704ea8556c39d1a84-head
- Installation option: Docker install
- Digital Ocean access token setup

<img width="591" alt="01-rancher-about" src="https://user-images.githubusercontent.com/40877149/137366774-e18da635-1337-4a7d-9f20-b3630b183e3f.png">

**Information about the Cluster**
- Kubernetes version: v1.21.5-rancher1-1
- Cluster Type: Downstream
   - RKE1
   - Digital Ocean
   - 3 node
   - Network Provider: Canal
 
   - If downstream, what type of cluster: Custom

**Describe the bug**

After creating a Workload -> Deployment with Ports Service Type set to Cluster IP. When trying to edit the configuration and change the Service Type from Cluster IP to Node Port the changes are **NOT** saved.There is no error presented in the UI but when going back in to edit the configuration the Service Type is back to Cluster IP and not saved as Node Port.

**To Reproduce**
 

1. Starting from the Rancher homepage `/dashboard/home`
2. Click the hamburger menu -> Cluster Management
3. Click `Explore` next to the cluster from the setup instructions
4. Click Workload -> Create -> Deployment
   - Namespace: default
   - Name: hello-world-test
   - Container Image: `brudnak/hello-rancher-golang`
5. Click `Add Port`
   - Service Type: Cluster IP
   - Private Container Port: 3000
   - Protocol: TCP
6. Click `Create`
7. You are then redirected back to the Deployments page
8. Click the `ellipsis` (three dots) next to the deployment that was just created -> Edit Config
9. Under the Ports section change the previously created Service Type from `Cluster IP` to `Node Port`
10. Under `Listening Port` enter -> `30456`
11. Click `Save`. You can see the network response in images labeled `A1` & `A2` in the Screenshots section below.
 
**Result**

Now if you repeat step `#8` from above, and look under the Ports section. You will see that the Service Type has **NOT** been updated to `Node Port` and that it is still set to `Cluster IP` even though we just changed it. And **DID NOT** receive any errors during the update/save process.

**Expected Result**

I would expect the Service Type to have changed from `Cluster IP` to `Node Port`. There were no errors indicating the save was not successful or that this was an invalid use or operation. Under the Screenshots section below there is a gif demonstrating this issue. 

**Screenshots**

![new-rancher-issuegif](https://user-images.githubusercontent.com/40877149/137366910-e2e7438b-3024-4eaf-a994-a637ff37de5c.gif)

<img width="935" alt="02-a1" src="https://user-images.githubusercontent.com/40877149/137370929-86fa91a0-2b43-42af-90c7-37346decbe5e.png">

<img width="968" alt="03-a2" src="https://user-images.githubusercontent.com/40877149/137370958-77754ccd-b4e3-42f1-bc1f-26513b1b19e3.png">

**Comment by brudnak** [comment]
### Reproduction Environment
#### Issue Persists in some test cases :bug:

| Component               | Type / Version                                     |
| ----------------------- | -------------------------------------------------- |
| Rancher version:        | v2.6-931a0f09d0f85ade0dcaa8fbe9d6e0c58ae52bbc-head |
| Rancher cluster type    | Single Node Docker Install                         |
| Docker version          | 20.10.7, build 20.10.7-0ubuntu1~20.04.2            |
| Helm version            | v2.16.8-rancher1                                   |
| Downstream cluster type | Custom RKE Linode                                  |
| Downstream K8s version  | v1.21.5-rancher1-1                                 |
| Cloud credentials       | Linode API Token setup prior to steps              |
| Browser Type            | Google Chrome                                      |
| Browser Version         | 95.0.4638.69 (Official Build) (x86_64)             |

>Thinking in terms of two scenarios here, one is solved. And one is still occurring.
> 1. :white_check_mark: User creates a deployment not giving the service-type a name
>    - The fix on clearing the name field on service-type selection solved this case
> 2. :x: If the user provides the same name for example "app1-service" for the previous Cluster IP and tries to reenter that same name for the Node Port it will not work
> 
> Another similar issue / use case has been found here: [https://github.com/rancher/dashboard/issues/4508](https://github.com/rancher/dashboard/issues/4508)
>
>You are able to keep the same name in the old UI.

## Reproduction steps

1. Starting from the **default** Rancher homepage `/dashboard/home`
    - With a downstream cluster created as described in the above `Validation Environment` table
2. Click `hamburger menu` -> `Cluster Management` -> `Explore` (next to cluster described in validation environment table) -> `Workload` -> `Deployments` -> `Create`
3. Fill out the deployment form:
    - Name: `hello-world`
    - Container Image: `brudnak/hello-rancher-golang`
    - Click `Add Port`
        - Service Type: `Cluster IP`
        - Name: `myapp-service`
        - Private Container Port: `3000`
        - Protocol: `TCP`
    - Click `Create` -> to create the deployment hello-world -> redirected back to the `Deployments` page -> wait for the deployment state to become `Active`
4. Click the `ellipsis (three dots)` next to the `hello-world` deployment -> click `Edit Config`
5. **Edit** the configuration like the following
    - Service Type: `Node Port`
    - Name: `myapp-service`
    - Private Container Port: `3000`
    - Listening Port: `30456`
6. Click `Save` -> there is a spinning "saving" animation on the "Save" button -> then redirected back to the `Deployments` page. The deployment has not changed to Node Port, and is still set as Cluster IP. See the video below, under "Screenshots & Video"
  
### Additional Info

There is a similar issue with trying to edit an existing Node Port here: [https://github.com/rancher/dashboard/issues/4508](https://github.com/rancher/dashboard/issues/4508)

### Screenshots & Video

https://user-images.githubusercontent.com/40877149/140266937-501ff557-4d95-4eb0-a5ed-0dea56b24bf6.mov

<hr>

**Comment by gaktive** [comment]
Hopefully whatever is fixed here also fixes #4508

**Comment by brudnak** [comment]
# My checks FAILED

## Reproduction Environment


#### Not required.

<hr>
  
## Validation Environment

**Occurs on both**

- `v2.6-44b8030a00b29f9f5354c645f3a90ede2eea53e0-head`
- `v2.6.3-rc2`

| Component               | Version / Type                                     |
| ----------------------- | -------------------------------------------------- |
| Rancher version         | v2.6.3-rc2                                         |
| Rancher version         | v2.6-44b8030a00b29f9f5354c645f3a90ede2eea53e0-head |
| Rancher cluster type    | Docker                                             |
| Docker version          | 20.10.7, build 20.10.7-0ubuntu5~20.04.2            |
| Helm version            | v2.16.8-rancher1                                   |
| Downstream cluster type | RKE1 Linode                                        |
| Downstream K8s version  | v1.21.6                                            |
| Browser type            | Google Chrome                                      |
| Browser version         | 96.0.4664.55 (Official Build) (x86_64)             |
| Cloud Credentials       | Linode cloud credentials setup prior to steps      |

![new-head](https://user-images.githubusercontent.com/40877149/143210437-a4cc2a93-eec3-4cf5-9fe3-758396299112.png)

![rc2](https://user-images.githubusercontent.com/40877149/143210468-30d4d8c6-4a50-4767-8fb1-cfadc08fe212.png)

## Validation steps

### ❌  Scenario 1 | Node Port | Blank Name | Change Listening Port

<hr>

### TLDR:

>Creating a deployment without a name, then deciding you want to change the listening port, creates multiple listening ports without the users knowledge because this is not being shown in the UI.

1. Starting from the **default** Rancher homepage `/dashboard/home`
2. Click `hamburger menu` -> click `Cluster Management` -> click `Explore` (next to cluster) -> click `Workload` -> click `Deployments` -> click `Create`
3. Fill out the deployment form
   - Namespace: `default`
   - Name: `hello-world`
   -  Container Image: `brudnak/hello-rancher-golang`
   -  Click `Add Port`
        - Service Type: `Node Port`
        - Name: leave **name** field blank
        - Private Container Port: `3000`
        - Listening Port: `30456`
   - Click `Create` -> to create the deployment
4. Deployment successfully working

> User decides they want to change the listening port from **30456** to **30442**


5. Click the `ellipsis` next to the `hello-world` deployment
6. Click `Edit Config`
7. Observe the port section while editing the config, the form isn't picking up previous selections
    - Service Type:
      - Was: `Node Port`
      - Now: `Do not create a service`
8. Change service type back to `Node Port`
9. Leave `Name` field blank
10. Enter newly desired Listening Port `30442`

**RESULT**

You are redirected back to the `Deployments` page. The `hello-world` deployment has remained in an `Active` state and is still showing endpoint `30456/TCP` instead of being updated to `30442/TCP`. However, if you change the URL to port `30442` you will also see the deployment.

If you change to a port we haven't entered yet like `30441` you'll get a page not found screen.

**1st VIDEO ATTACHED BELOW**

https://user-images.githubusercontent.com/40877149/143210183-b0c60c40-322e-46e2-acc6-8cd114f55a39.mov

<hr>

### ❌  Scenario 2 | Named Cluster IP | Change to Node Port with Same Name

### TLDR:

>When creating a Cluster IP with a name, then changing to a Node Port with the same name. The deployment does not change to a Node Port service type

1. Starting from the **default** Rancher homepage `/dashboard/home`
2. Click `hamburger menu` -> click `Cluster Management` -> click `Explore` (next to cluster) -> click `Workload` -> click `Deployments` -> click `Create`
3. Fill out the deployment form
   - Namespace: `default`
   - Name: `hello-world`
   -  Container Image: `brudnak/hello-rancher-golang`
   -  Click `Add Port`
        - Service Type: `Cluster IP`
        - Name: `my-port-5`
        - Private Container Port: `3000`
   - Click `Create` -> to create the deployment
4. Deployment successfully working

> User selected Cluster IP by accident and actually wants a Node Port, so they go back into edit config to change it

5. Click the `ellipsis` next to the deployment
6. Click `Edit Config`
7. Change the Service Type from Cluster IP to Node Port
8. Leave the same name `my-port-5`
9. Click `Save`

**RESULT**

The service type does not change from Cluster IP to Node Port.

**2nd VIDEO ATTACHED BELOW**

https://user-images.githubusercontent.com/40877149/143210288-cf8cc8c1-5ec5-4460-8c6b-c120a2508a67.mov

<hr>

**Comment by brudnak** [comment]
# My checks FAILED

<hr>
  
## Validation Environment

| Component               | Version / Type                                     |
| ----------------------- | -------------------------------------------------- |
| Rancher version         | v2.6-4375161a2780765f0fd58af2936876a3e0254741-head |
| Rancher cluster type    | Docker                                             |
| Docker version          | 20.10.7, build 20.10.7-0ubuntu5~20.04.2            |
| Helm version            | v2.16.8-rancher1                                   |
| Downstream cluster type | RKE1 Linode                                        |
| Downstream K8s version  | v1.21.6-rancher1-1                                 |
| Browser type            | Google Chrome                                      |
| Browser version         | 96.0.4664.55 (Official Build) (x86_64)             |

## Validation steps

### ❌  Scenario 1 | Node Port | Blank Name | Change Listening Port

<hr>

### TLDR:

1. Starting from the **default** Rancher homepage `/dashboard/home`
2. Click `hamburger menu` -> click `Cluster Management` -> click `Explore` (next to cluster) -> click `Workload` -> click `Deployments` -> click `Create`
3. Fill out the deployment form
   - Namespace: `default`
   - Name: `hello-world`
   -  Container Image: `brudnak/hello-rancher-golang`
   -  Click `Add Port`
        - Service Type: `Node Port`
        - Name: leave **name** field blank
        - Private Container Port: `3000`
        - Listening Port: `30441`
   - Click `Create` -> to create the deployment
4. Deployment successfully working

> User decides they want to change the listening port from **30441** to **30442**


5. Click the `ellipsis` next to the `hello-world` deployment
6. Click `Edit Config`
7. Observe the port section while editing the config, the form isn't picking up previous selections
    - Service Type:
      - Was: `Node Port`
      - Now: `Do not create a service`
8. Change service type back to `Node Port`
9. Leave `Name` as is, it should be a generated name
10. Enter newly desired Listening Port `30442`

**----------------------------------------------------------------------------------------------**

**INITIAL RESULT**

You are redirected back to the deployments page. The hello-world deployment remains active,  the `Endpoints` link does not get updated. It still showing `30441/TCP`. If you click this link you get page not found, however if you manually change the URL to port `30442` you can see the deployment.

**----------------------------------------------------------------------------------------------**

11. Edit the config again as in steps `6 through 10` and change the listening port to `30443`. 
12. You'll see the same results, where the UI is still showing `30441/TCP` for the `Endpoints` link. Even though we've changed it twice now to `30442` and now `30443`. If you change the URL to port 

**RESULT**

The `Endpoints` URL is not being updated on the `Deployments` page. Even though it's actually being changed if you navigate to the correct URL. See the screen shot and video listed below. Screen shot labeled `A1`

**SCREENSHOT**

![Screen Shot 2021-11-29 at 11 28 52 AM](https://user-images.githubusercontent.com/40877149/143999372-74bb56e2-0095-4c9e-b2b1-cfe263e583eb.png)

**VIDEO**

https://user-images.githubusercontent.com/40877149/143999420-191b5d10-bcc6-4c99-9644-931d692f3854.mov

<hr>

### ❌  Scenario 2 | Named Cluster IP | Change to Node Port with Same Name

### TLDR:

>When creating a Cluster IP with a name, then changing to a Node Port with the same name. The deployment does not change to a Node Port service type

1. Starting from the **default** Rancher homepage `/dashboard/home`
2. Click `hamburger menu` -> click `Cluster Management` -> click `Explore` (next to cluster) -> click `Workload` -> click `Deployments` -> click `Create`
3. Fill out the deployment form
   - Namespace: `default`
   - Name: `hello-world`
   -  Container Image: `brudnak/hello-rancher-golang`
   -  Click `Add Port`
        - Service Type: `Cluster IP`
        - Name: `my-port`
        - Private Container Port: `3000`
   - Click `Create` -> to create the deployment
4. Deployment successfully working

> User selected Cluster IP by accident and actually wants a Node Port, so they go back into edit config to change it

5. Click the `ellipsis` next to the deployment
6. Click `Edit Config`
7. Change the Service Type from Cluster IP to Node Port
8. Set `Listening Port` to `30441`
9. Leave the same name `my-port`
10. Click `Save`

**RESULT**

The service type does not change from Cluster IP to Node Port.

**VIDEO**

https://user-images.githubusercontent.com/40877149/143999484-17f174d8-409a-456b-9ca7-ece3d070204d.mov

<hr>

**Comment by mantis-toboggan-md** [comment]
@brudnak after poking around the UI, I've filed a backend issue for scenario 1 here https://github.com/rancher/rancher/issues/35690 -- scenario 2 does look like something we (still :( ) need to fix on the UI side.

**Comment by brudnak** [comment]
> @brudnak after poking around the UI, I've filed a backend issue for scenario 1 here [rancher/rancher#35690](https://github.com/rancher/rancher/issues/35690) -- scenario 2 does look like something we (still :( ) need to fix on the UI side.

@mantis-toboggan-md thanks for the info 😄

**Comment by mantis-toboggan-md** [comment]
I've done a little more testing and it looks like both scenario 1 and 2 are actually caused by the same backend issue. When I change a workload service type from clusterip to nodeport, I do see the nodeport service created as expected, but the workload table isn't updating. Previously there was a UI issue here, where changing a port from clusterip to nodeport did _not_ create a nodeport service, but that appears to be resolved. As far as I can tell the problem with the table not updating is distinct; that column is based off annotation the UI never sets/updates itself.

**Comment by brudnak** [comment]
Waiting for backend issue to complete checks

**Comment by brudnak** [comment]
# My checks PASSED
 
## Reproduction Environment


#### Not required.

<hr>
  
## Validation Environment

| Component               | Version / Type                                     |
| ----------------------- | -------------------------------------------------- |
| Rancher version         | v2.6-fd930f76545c64fdad8b685d86090a8fbb8bebcf-head |
| Rancher cluster type    | Docker                                             |
| Docker version          | 20.10.7, build 20.10.7-0ubuntu5~20.04.2            |
| Helm version            | v2.16.8-rancher1                                   |
| Downstream cluster type | RKE1 Linode                                        |
| Downstream K8s version  | v1.21.7-rancher1-1                                 |
| Browser type            | Opera GX                                           |
| Browser version         | LVL3 (core: 82.0.4227.25) (x86_64)                 |

## Validation steps

1. Create deployment
    - Image used: `brudnak/hello-rancher-golang`
    - Ports:
        - Service Type: `Cluster IP`
        - Name: `app1`
        - Private Container Port: `3000`
2. After the deployment is created. Edit the config of the deployment
3. Edited deployment details
    - Service Type: `Node Port`
    - Name: `app1` (leave the same name for port)
    - Listening Port: `30339`
4. Save the deployment

**RESULT**

The deployment successfully gets converted from a Cluster IP to a Node Port service type.
